File: apps/analytics/insights/frequency.py
"""Session frequency insight.

Computes sessions/week, current streak, and consistency percentage.
"""

from __future__ import annotations

from collections import defaultdict
from datetime import datetime, timedelta

from queries import WorkoutRecord


def compute(records: list[WorkoutRecord]) -> dict | None:
    """Return a frequency payload or None if insufficient data."""
    # Collect unique workout dates (one session per calendar date max)
    dates: set[str] = set()
    for r in records:
        if r.recorded_at:
            try:
                dt = datetime.fromisoformat(r.recorded_at.replace("Z", "+00:00"))
                dates.add(dt.date().isoformat())
            except ValueError:
                pass

    if len(dates) < 3:
        return None

    sorted_dates = sorted(dates)
    first = datetime.fromisoformat(sorted_dates[0]).date()
    last = datetime.fromisoformat(sorted_dates[-1]).date()
    total_weeks = max(1, (last - first).days / 7)

    sessions_per_week = round(len(sorted_dates) / total_weeks, 2)

    streak = _current_streak(sorted_dates)
    consistency_pct = _consistency_pct(sorted_dates, first, last)

    return {
        "sessionsPerWeek": sessions_per_week,
        "currentStreak": streak,
        "consistencyPct": consistency_pct,
        "totalSessions": len(sorted_dates),
    }


def _current_streak(sorted_date_strs: list[str]) -> int:
    """Count consecutive days from the most recent session backwards."""
    if not sorted_date_strs:
        return 0
    dates = [datetime.fromisoformat(d).date() for d in sorted_date_strs]
    dates_set = set(dates)
    streak = 0
    cursor = dates[-1]
    while cursor in dates_set:
        streak += 1
        cursor -= timedelta(days=1)
    return streak


def _consistency_pct(sorted_date_strs: list[str], first, last) -> float:
    total_weeks = max(1, ((last - first).days + 1) // 7)
    if total_weeks < 1:
        return 100.0

    weeks_with_session: set[str] = set()
    for d in sorted_date_strs:
        dt = datetime.fromisoformat(d).date()
        iso = dt.isocalendar()
        weeks_with_session.add(f"{iso.year}-W{iso.week:02d}")

    return round(min(100.0, len(weeks_with_session) / total_weeks * 100), 1)

File: apps/backend/api/src/analytics/pipelines/frequency.ts
/**
 * Session-frequency insight (aggregate, exercise_id = null).
 *
 * Ports apps/backend/analytics/insights/frequency.py. Computes sessions/week,
 * the current consecutive-day streak, a weekly consistency percentage, and a
 * trailing window of workout dates for heatmap rendering.
 */

import type { WorkoutRecord } from '../record';
import { isoWeekKeyFromTimestamp } from '../iso-week';
import { wallClockDateKey, daysBetweenDateKeys, shiftDateKey } from './time';
import { pyRound } from './round';

export interface FrequencyPayload {
  readonly sessionsPerWeek: number;
  readonly currentStreak: number;
  readonly consistencyPct: number;
  readonly totalSessions: number;
  readonly workoutDates: string[];
}

export function computeFrequency(records: readonly WorkoutRecord[]): FrequencyPayload | null {
  // Collect unique workout dates (one session per calendar date max).
  const dates = new Set<string>();
  for (const r of records) {
    if (r.recordedAt === null) continue;
    const key = wallClockDateKey(r.recordedAt);
    if (key !== null) dates.add(key);
  }

  if (dates.size < 3) return null;

  const sortedDates = [...dates].sort();
  const first = sortedDates[0];
  const last = sortedDates[sortedDates.length - 1];
  const totalWeeks = Math.max(1, daysBetweenDateKeys(first, last) / 7);

  const sessionsPerWeek = pyRound(sortedDates.length / totalWeeks, 2);
  const streak = currentStreak(sortedDates);
  const consistencyPct = consistencyPercentage(sortedDates, first, last);

  // Include the last 28 days of workout dates for heatmap rendering.
  const workoutDates = sortedDates.length > 28 ? sortedDates.slice(-28) : sortedDates;

  return {
    sessionsPerWeek,
    currentStreak: streak,
    consistencyPct,
    totalSessions: sortedDates.length,
    workoutDates,
  };
}

/** Consecutive calendar days from the most recent session backwards. */
function currentStreak(sortedDateKeys: readonly string[]): number {
  if (sortedDateKeys.length === 0) return 0;
  const present = new Set(sortedDateKeys);
  let streak = 0;
  let cursor = sortedDateKeys[sortedDateKeys.length - 1];
  while (present.has(cursor)) {
    streak += 1;
    cursor = shiftDateKey(cursor, -1);
  }
  return streak;
}

/** Percentage of elapsed weeks that contain at least one session, capped at
 * 100 and rounded to one decimal, matching frequency.py:_consistency_pct. */
function consistencyPercentage(
  sortedDateKeys: readonly string[],
  first: string,
  last: string
): number {
  const totalWeeks = Math.max(1, Math.floor((daysBetweenDateKeys(first, last) + 1) / 7));

  const weeksWithSession = new Set<string>();
  for (const key of sortedDateKeys) {
    const weekKey = isoWeekKeyFromTimestamp(key);
    if (weekKey !== null) weeksWithSession.add(weekKey);
  }

  return pyRound(Math.min(100, (weeksWithSession.size / totalWeeks) * 100), 1);
}
